### PR TITLE
[BTFS-1044]feat: new grpc conn

### PR DESCRIPTION
**Pull request by laipogo**

### CODE REVIEW

File: protos/helpers/client.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+func NewGRPCConn(ctx context.Context, address string, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
+	u, err := parse(address)
+	if err != nil {
+		return nil, err
+	}
+	if u.schema == "http" {
+		opts = append(opts, grpc.WithInsecure())
+	} else if u.schema == "https" {
+		h2creds := credentials.NewTLS(&tls.Config{NextProtos: []string{"h2"}})
```
Comment (taiyangc) [inline protos/helpers/client.go]: I have been using `credentials.NewTLS(&tls.Config{})` since "h2" is defaultly enabled on tls config

Comment (laipogo) [reply protos/helpers/client.go]: oh, i see. i just forgot your comments in the ticket.

File: protos/helpers/client.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+		opts = append(opts, grpc.WithTransportCredentials(h2creds))
+	} else {
+		return nil, errors.New("not supported schema")
+	}
+	return grpc.DialContext(ctx, fmt.Sprintf("%s:%d", u.host, u.port), opts...)
+}
+
+func parse(rawU string) (*parsedURL, error) {
```
Comment (taiyangc) [inline protos/helpers/client.go]: Why not use https://golang.org/pkg/net/url/#Parse? I think it's standard

Comment (laipogo) [reply protos/helpers/client.go]: > Why not use https://golang.org/pkg/net/url/#Parse? I think it's standard

actually i used it, in line 54.
but it need more things to todo.
e.x. http://localhost:8080 will be parsed to `{schema: "http", host: "localhost:8080",port:"8080"}`

Comment (taiyangc) [reply protos/helpers/client.go]: ok

File: protos/helpers/client.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		rawU = defaultSchema + ":" + rawU
+	}
+	if !strings.Contains(rawU, "://") {
+		rawU = defaultSchema + "://" + rawU
+	}
+	u, err := url.Parse(rawU)
+	if err != nil {
+		return nil, err
+	}
+	h, p, err := splitHostPort(u)
```
Comment (taiyangc) [inline protos/helpers/client.go]: Just found https://golang.org/pkg/net/url/#URL.Port, i think you can safely use url functions, which is more stable.

Comment (laipogo) [reply protos/helpers/client.go]: will do it.

Comment (laipogo) [reply protos/helpers/client.go]: url library doesn’t handle schema and port == "", and doesn’t check validaity of host. so i add some logic to do them.

**Comment by taiyangc** [comment]
@wxue @jin-tron @JustinGaoF please help with another review, thanks. You will use this client sometime during your projects

**Comment by laipogo** [comment]
> code lgtm, but this shouldn't put into protos/ since it's not proto files;
> should put into utils maybe?

done

**Pull request merged**